File: src/main/java/me/testaccount666/serversystem/ServerSystem.java
package me.testaccount666.serversystem;

import lombok.Getter;
import lombok.SneakyThrows;
import me.testaccount666.migration.LegacyDataMigrator;
import me.testaccount666.migration.essentials.EssentialsMigrator;
import me.testaccount666.serversystem.clickablesigns.SignManager;
import me.testaccount666.serversystem.commands.executables.kit.manager.KitManager;
import me.testaccount666.serversystem.commands.executables.warp.manager.WarpManager;
import me.testaccount666.serversystem.commands.management.CommandManager;
import me.testaccount666.serversystem.commands.management.CommandReplacer;
import me.testaccount666.serversystem.listener.management.ListenerManager;
import me.testaccount666.serversystem.managers.config.ConfigurationManager;
import me.testaccount666.serversystem.managers.database.economy.AbstractEconomyDatabaseManager;
import me.testaccount666.serversystem.managers.database.economy.MySqlEconomyDatabaseManager;
import me.testaccount666.serversystem.managers.database.economy.SqliteEconomyDatabaseManager;
import me.testaccount666.serversystem.managers.database.moderation.AbstractModerationDatabaseManager;
import me.testaccount666.serversystem.managers.database.moderation.MySqlModerationDatabaseManager;
import me.testaccount666.serversystem.managers.database.moderation.SqliteModerationDatabaseManager;
import me.testaccount666.serversystem.placeholderapi.PlaceholderApiSupport;
import me.testaccount666.serversystem.placeholderapi.PlaceholderManager;
import me.testaccount666.serversystem.updates.UpdateManager;
import me.testaccount666.serversystem.userdata.CachedUser;
import me.testaccount666.serversystem.userdata.OfflineUser;
import me.testaccount666.serversystem.userdata.UserManager;
import me.testaccount666.serversystem.userdata.money.EconomyProvider;
import me.testaccount666.serversystem.userdata.money.vault.EconomyVaultAPI;
import me.testaccount666.serversystem.utils.Version;
import me.testaccount666.serversystem.utils.VersionInfo;
import org.bukkit.Bukkit;
import org.bukkit.configuration.file.YamlConfiguration;
import org.bukkit.plugin.java.JavaPlugin;

import java.io.File;
import java.io.IOException;
import java.nio.file.Path;
import java.util.logging.Level;
import java.util.logging.Logger;

public final class ServerSystem extends JavaPlugin {
    public static final Version MINIMUM_MINECRAFT_VERSION = new Version("1.21");
    public static final Version CURRENT_VERSION = new Version(VersionInfo.CLEAN_VERSION);
    public static ServerSystem Instance;
    @Getter
    private static Logger _Log;
    @Getter
    private final ServiceRegistry _registry = new ServiceRegistry();

    public static Version getServerVersion() {
        var version = Bukkit.getMinecraftVersion();

        getLog().log(Level.FINE, "Server version: ${version}");

        return new Version(version);
    }

    @SneakyThrows
    @Override
    public void onEnable() {
        Instance = this;
        _Log = getLogger();

        var minecraftVersion = getServerVersion();
        if (minecraftVersion.compareTo(MINIMUM_MINECRAFT_VERSION) < 0)
            getLog().warning("You're running an unsupported/legacy version of Minecraft! " +
                    "Please update to at least ${MINIMUM_MINECRAFT_VERSION.getVersion()}!");

        var migrator = new LegacyDataMigrator();
        if (migrator.isLegacyDataPresent()) {
            getLog().log(Level.INFO, "Legacy data detected. Attempting to migrate...");
            migrator.prepareMigration();
        }

        var previousVersionFile = new File(getDataFolder(), "previousVersion.yml");
        var previousVersionConfig = YamlConfiguration.loadConfiguration(previousVersionFile);
        previousVersionConfig.set("previousVersion", CURRENT_VERSION.toString());
        try {
            previousVersionConfig.save(previousVersionFile);
        } catch (IOException exception) {
            throw new RuntimeException("Error updating 'previousVersion'", exception);
        }

        getRegistry().registerService(EssentialsMigrator.class, new EssentialsMigrator());

        try {
            initialize();
        } catch (Exception exception) {
            getLog().log(Level.SEVERE, "Failed to initialize the plugin", exception);
            Bukkit.getPluginManager().disablePlugin(this);
            return;
        }

        var configManager = getRegistry().getService(ConfigurationManager.class);
        getRegistry().registerService(UpdateManager.class, new UpdateManager(this, configManager)).start();

        Bukkit.getScheduler().runTaskLater(this, migrator::migrateLegacyData, 1L);

        Bukkit.getScheduler().runTaskLater(this, () -> {
            var registry = getRegistry();

            registry.getServiceOptional(CommandManager.class).ifPresent(CommandManager::registerCommands);
            registry.getServiceOptional(ListenerManager.class).ifPresent(ListenerManager::registerListeners);
            registry.getServiceOptional(PlaceholderManager.class).ifPresent(PlaceholderManager::registerPlaceholders);
            registry.getServiceOptional(CommandReplacer.class).ifPresent(CommandReplacer::replaceCommands);

            //var essentialsMigrator = new EssentialsMigrator();
            //if (essentialsMigrator.isEssentialsInstalled()) essentialsMigrator.migrateTo();
        }, 2);
    }

    private void initialize() throws Exception {
        var registry = getRegistry();

        var configManager = new ConfigurationManager(this);
        registry.registerService(ConfigurationManager.class, configManager).loadAllConfigs();

        var moderationType = configManager.getModerationConfig().getString("Moderation.StorageType.Value");
        var moderationDbManager = switch (moderationType.toLowerCase()) {
            case "sqlite" -> new SqliteModerationDatabaseManager(getDataFolder());
            case "mysql" -> new MySqlModerationDatabaseManager(configManager.getModerationConfig());
            default -> throw new IllegalStateException("Unsupported moderation storage: ${moderationType} - Supported values: sqlite, mysql");
        };
        registry.registerService(AbstractModerationDatabaseManager.class, moderationDbManager);

        var essentialsMigrator = registry.getService(EssentialsMigrator.class);
        if (!essentialsMigrator.isEssentialsInstalled() && EconomyVaultAPI.isVaultInstalled()) EconomyVaultAPI.initialize();

        PlaceholderApiSupport.registerPlaceholders();

        moderationDbManager.initialize();

        var economyStorageType = configManager.getEconomyConfig().getString("Economy.StorageType.Value");
        var economyDbManager = switch (economyStorageType.toLowerCase()) {
            case "sqlite" -> new SqliteEconomyDatabaseManager(getDataFolder());
            case "mysql" -> new MySqlEconomyDatabaseManager(configManager.getEconomyConfig());
            default -> throw new IllegalStateException("Unsupported economy storage: ${economyStorageType} - Supported values: sqlite, mysql");
        };
        registry.registerService(AbstractEconomyDatabaseManager.class, economyDbManager);

        var commandManager = new CommandManager(configManager.getCommandsConfig());
        registry.registerService(CommandManager.class, commandManager);
        registry.registerService(ListenerManager.class, new ListenerManager(commandManager));
        registry.registerService(PlaceholderManager.class, new PlaceholderManager());
        registry.registerService(EconomyProvider.class, new EconomyProvider(configManager.getEconomyConfig()));
        registry.registerService(UserManager.class, new UserManager());

        registry.registerService(KitManager.class, new KitManager());

        registry.registerService(CommandReplacer.class, new CommandReplacer());

        Bukkit.getScheduler().runTask(this, () -> {
            var warpFile = Path.of(getDataFolder().getPath(), "data", "warps.yml").toFile();
            var warpConfig = YamlConfiguration.loadConfiguration(warpFile);

            registry.registerService(WarpManager.class, new WarpManager(warpConfig, warpFile));
            registry.registerService(SignManager.class, new SignManager()).loadSignTypes();
        });
    }

    @Override
    public void onDisable() {
        var registry = getRegistry();

        registry.getServiceOptional(UserManager.class).ifPresent(this::saveAllUsers);

        registry.getServiceOptional(CommandManager.class).ifPresent(CommandManager::unregisterCommands);
        registry.getServiceOptional(ListenerManager.class).ifPresent(ListenerManager::unregisterListeners);

        PlaceholderApiSupport.unregisterPlaceholders();

        registry.getServiceOptional(AbstractEconomyDatabaseManager.class).ifPresent(AbstractEconomyDatabaseManager::shutdown);
        registry.getServiceOptional(AbstractModerationDatabaseManager.class).ifPresent(AbstractModerationDatabaseManager::shutdown);

        registry.clearServices();
    }

    private void saveAllUsers(UserManager userManager) {
        userManager.getCachedUsers().stream().map(CachedUser::getOfflineUser).forEach(OfflineUser::save);
    }
}

File: src/main/kotlin/me/testaccount666/serversystem/ServerSystem.kt
package me.testaccount666.serversystem

import me.testaccount666.migration.LegacyDataMigrator
import me.testaccount666.migration.plugins.MigratorRegistry
import me.testaccount666.paperktx.PaperKTX
import me.testaccount666.paperktx.scheduler.skedule.okkero.schedule
import me.testaccount666.serversystem.clickablesigns.SignManager
import me.testaccount666.serversystem.commands.executables.kit.manager.KitManager
import me.testaccount666.serversystem.commands.executables.waypoints.warp.manager.WarpManager
import me.testaccount666.serversystem.commands.management.CommandManager
import me.testaccount666.serversystem.commands.management.CommandReplacer
import me.testaccount666.serversystem.listener.management.ListenerManager
import me.testaccount666.serversystem.managers.config.ConfigurationManager
import me.testaccount666.serversystem.managers.database.economy.*
import me.testaccount666.serversystem.managers.database.moderation.*
import me.testaccount666.serversystem.placeholderapi.PlaceholderApiSupport
import me.testaccount666.serversystem.placeholderapi.PlaceholderManager
import me.testaccount666.serversystem.updates.UpdateManager
import me.testaccount666.serversystem.userdata.UserManager
import me.testaccount666.serversystem.userdata.money.EconomyProvider
import me.testaccount666.serversystem.userdata.money.vault.EconomyVaultAPI
import me.testaccount666.serversystem.utils.Version
import me.testaccount666.serversystem.utils.VersionInfo
import org.bukkit.Bukkit
import org.bukkit.configuration.file.YamlConfiguration
import org.bukkit.plugin.java.JavaPlugin
import java.io.File
import java.io.IOException
import java.nio.file.Path
import java.util.logging.Level
import java.util.logging.Logger

class ServerSystem : JavaPlugin() {
    val registry = ServiceRegistry()

    override fun onLoad() {
        instance = this
        log = logger

        if (serverVersion < MINIMUM_MINECRAFT_VERSION) {
            log.warning(
                "You're running an unsupported/legacy version of Minecraft! " +
                        "Please update to at least ${MINIMUM_MINECRAFT_VERSION.version}!"
            )
        }
    }

    override fun onEnable() {
        PaperKTX.plugin = this
        val migrator = LegacyDataMigrator()
        if (migrator.isLegacyDataPresent) {
            log.log(Level.INFO, "Legacy data detected. Attempting to migrate...")
            migrator.prepareMigration()
        }

        val previousVersionFile = File(dataFolder, "previousVersion.yml")
        val previousVersionConfig = YamlConfiguration.loadConfiguration(previousVersionFile)
        previousVersionConfig["previousVersion"] = CURRENT_VERSION.toString()
        try {
            previousVersionConfig.save(previousVersionFile)
        } catch (exception: IOException) {
            throw RuntimeException("Error updating 'previousVersion'", exception)
        }

        try {
            initialize()
        } catch (exception: Exception) {
            log.log(Level.SEVERE, "Failed to initialize the plugin", exception)
            Bukkit.getPluginManager().disablePlugin(this)
            return
        }

        val configManager = registry.getService<ConfigurationManager>()
        registry.registerService(UpdateManager(this, configManager)).start()

        schedule {
            waitFor(1)
            migrator.migrateLegacyData()
        }

        schedule {
            waitFor(2)
            registry.getServiceOrNull<CommandManager>()?.registerCommands()
            registry.getServiceOrNull<ListenerManager>()?.registerListeners()
            registry.getServiceOrNull<PlaceholderManager>()?.registerPlaceholders()
            registry.getServiceOrNull<CommandReplacer>()?.replaceCommands()
        }
    }

    private fun initialize() {
        val configManager = ConfigurationManager(this)
        registry.registerService(configManager).loadAllConfigs()

        val moderationType = configManager.moderationConfig.getString("Moderation.StorageType.Value")
        requireNotNull(moderationType) { "Moderation storage type not set?!" }

        val moderationDbManager =
            when (moderationType.lowercase()) {
                "sqlite" -> SqliteModerationDatabaseManager(dataFolder)
                "mysql" -> MySqlModerationDatabaseManager(configManager.moderationConfig)
                else -> error("Unsupported moderation storage: $moderationType - Supported values: sqlite, mysql")
            }
        registry.registerService<ModerationDatabaseManager>(moderationDbManager)

        val migratorRegistry = registry.registerService(MigratorRegistry()).also(MigratorRegistry::registerMigrators)
        PlaceholderApiSupport.registerPlaceholders()

        moderationDbManager.initialize()

        val economyStorageType = configManager.economyConfig.getString("Economy.StorageType.Value")
        requireNotNull(economyStorageType) { "Economy storage type not set?!" }

        val economyDbManager =
            when (economyStorageType.lowercase()) {
                "sqlite" -> SqliteEconomyDatabaseManager(dataFolder)
                "mysql" -> MySqlEconomyDatabaseManager(configManager.economyConfig)
                else -> error("Unsupported economy storage: $economyStorageType - Supported values: sqlite, mysql")
            }
        registry.registerService<EconomyDatabaseManager>(economyDbManager)

        val commandManager = CommandManager(configManager.commandsConfig)
        registry.registerService(commandManager)
        registry.registerService(ListenerManager(commandManager))
        registry.registerService(PlaceholderManager())
        registry.registerService(EconomyProvider(configManager.economyConfig))
        registry.registerService(UserManager())

        registry.registerService(KitManager())

        registry.registerService(CommandReplacer())

        val migratorCount = migratorRegistry.migrators.size

        if (migratorCount == 0 && configManager.economyConfig.getBoolean("Economy.HookIntoVault")) {
            if (EconomyVaultAPI.isVaultInstalled) EconomyVaultAPI.initialize()
        }

        if (migratorCount > 0) log.info("Not hooking into Vault since we found data migrators!")

        schedule {
            val warpFile = Path.of(dataFolder.path, "data", "warps.yml").toFile()
            val warpConfig = YamlConfiguration.loadConfiguration(warpFile)

            registry.registerService(WarpManager(warpConfig, warpFile))
            registry.registerService(SignManager()).loadSignTypes()
        }
    }

    override fun onDisable() = with(registry) {
        getServiceOrNull<UserManager>()?.run(::saveAllUsers)

        getServiceOrNull<CommandManager>()?.run(CommandManager::unregisterCommands)
        getServiceOrNull<ListenerManager>()?.run(ListenerManager::unregisterListeners)

        PlaceholderApiSupport.unregisterPlaceholders()

        getServiceOrNull<EconomyDatabaseManager>()?.shutdown()
        getServiceOrNull<ModerationDatabaseManager>()?.shutdown()

        clearServices()
    }

    private fun saveAllUsers(userManager: UserManager) = userManager.cachedUsers.forEach { it.offlineUser.save() }

    companion object {
        val MINIMUM_MINECRAFT_VERSION = Version("1.21")

        @JvmField
        val CURRENT_VERSION = Version(VersionInfo.CLEAN_VERSION)

        @JvmStatic
        lateinit var instance: ServerSystem

        @JvmStatic
        lateinit var log: Logger

        val serverVersion by lazy {
            val version = Bukkit.getMinecraftVersion()

            log.log(Level.FINE, "Server version: $version")

            return@lazy Version(version)
        }
    }
}
